Fix local volumes with common GID, and other fixes

svenssonaxel opened the pull request:
My suggested fixes to https://github.com/logicalclocks/rondb-docker/pull/8

svenssonaxel committed:
Fix permissions and use local mounts in CI

- Fix permission problems by
  - Letting the group have the same permissions as the owner for all
    directories and most files under `./autogenerated_files/`.
  - At run-time, make the container user's primary group GID match
    that of the host user.
- Add --volumes-in-local-dir to CI run.
- Rename `fix_user.sh` to `entrypoint.sh`.
- Refactor `build_run_docker.sh`.
- Minor fixes.
- Remove README section about this solved problem.

svenssonaxel committed:
Preparation for group ownership solution

- Revert most of "56aba32 Removed unnecessary ENV_FIELD; Creating volume list also if only VOLUME_TYPE_BENCH_DIRS==docker"
- Revert much of "c8b337b Removed debug statements", note that this
  commit contains other things than debug statemens.
- Revert much of "271bd25 Added flag --bench-dirs-in-volumes"
- Minor fixes.

svenssonaxel committed:
Fix entrypoint modes

svenssonaxel committed:
Increase ndb-wait-setup to 15 min

svenssonaxel committed:
Fix workflow

svenssonaxel committed:
Lint shell scripts

Fix most errors and warnings produced by:
shellcheck *.sh *.env resources/entrypoints/*

olapiv commented on .github/workflows/test_and_deploy.yml:
```diff
@@ -33,10 +33,10 @@ jobs:
             --num-api-nodes 1 \
             --run-benchmark sysbench_single \
-            --bench-dirs-in-volumes \
+            --volumes-in-local-dir \
```
Bit of a formality, but if the API gets changed, you can run build_run_docker.sh with `--save-sample-files`, to fix the sample files. They're just for links in blog posts and documentation. Or you just revert the changes that I had made in those files.

svenssonaxel replied:
Fixed, will push soon

olapiv commented on resources/config_templates/my.cnf:
```diff
@@ -81,7 +81,7 @@ ndb-batch-size=1M
 # set connectstring to ndb management server (used by all executables)
 ndb-connectstring=%s
-# ndb-wait-setup=10800
+ndb-wait-setup=900
```
What's this for?

svenssonaxel replied:
On my machine, sometimes, `mysqld_X` would fail, complaining that data nodes weren't ready in `30s` and suggested increasing this value.

olapiv commented on resources/entrypoints/entrypoint.sh:
```diff
@@ -0,0 +1,26 @@
+if [ "$(getent group "$HOST_GROUP_ID")" ]; then
+	echo "group $HOST_GROUP_ID exists."
+else
+	echo "group $HOST_GROUP_ID does not exist."
```
Just append `[Entrypoint]` to the echos for consistency.

svenssonaxel replied:
Fixed

olapiv commented on resources/entrypoints/mysqld.sh:
```diff
@@ -148,30 +150,30 @@ if [ "$MYSQL_USER" ]; then
-mysqladmin -uroot --password=$MYSQL_ROOT_PASSWORD shutdown --socket="$SOCKET"
-echo "[Entrypoint] Successfully shut down MySQLd"
+echo '[entrypoints/mysqld.sh] Shutting down MySQLd via mysqladmin...'
+mysqladmin -uroot --password="$MYSQL_ROOT_PASSWORD" shutdown --socket="$SOCKET"
```
There a lot of these changes - I guess it's best practice? In case the variable has a whitespace stored?

svenssonaxel replied:
Yes, most of these are part of commit `Lint shell scripts`. I used `shellcheck`, a linter for bash and other shells, to find those.

olapiv commented on resources/entrypoints/main.sh:
```diff
@@ -40,13 +57,38 @@ else
-		echo "[Entrypoint] Starting ndb_mgmd"
-		set -- "$@" -f $RONDB_DATA_DIR/config.ini --configdir=$RONDB_DATA_DIR/log
+		echo "[entrypoints/main.sh] Starting ndb_mgmd"
+		set -- "$@" -f "$RONDB_DATA_DIR/config.ini" --configdir="$RONDB_DATA_DIR/log"
```
What about the permissions of the logs/data that the mgmd creates?

svenssonaxel replied:
The `umask` value of `0002` gets inherited by all processes started from this script, also `ndb_mgmd`, see PID `31` in this example:
```
docker exec mgmd_1 bash -c 'ps -fuxa; grep -H Umask /proc/*/status'
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root          58  0.0  0.0   4360  3300 ?        Ss   07:04   0:00 bash -c ps -fuxa; grep -H Umask /proc/*/status
root          64  0.0  0.0   7060  1564 ?        R    07:04   0:00  \_ ps -fuxa
root           1  0.0  0.0   4360  3304 ?        Ss   07:02   0:00 /bin/bash ./docker_entrypoints/rondb_standalone/entrypoint.sh ndb_mgmd --ndb-nodeid=65 --initial
root          30  0.0  0.0   7236  4392 ?        S    07:02   0:00 sudo -E -u mysql /srv/hops/docker_entrypoints/rondb_standalone/main.sh ndb_mgmd --ndb-nodeid=65 --initial
mysql         31  0.8  0.0 1035104 25040 ?       Sl   07:02   0:01  \_ ndb_mgmd --ndb-nodeid=65 --initial --nodaemon -f /srv/hops/mysql-cluster/config.ini --configdir=/srv/hops/mysql-cluster/log
/proc/1/status:Umask:   0022
/proc/30/status:Umask:  0022
/proc/31/status:Umask:  0002
/proc/58/status:Umask:  0022
/proc/self/status:Umask:        0022
/proc/thread-self/status:Umask: 0022
```
Never had a problem with mgmd in testing. It appears that `ndb_mgmd` either does not create any files in the volumes or uses system default permissions, which will result in `0666 & ~UMASK = 0664` for files and `0777 & ~UMASK = 0775` for directories.

olapiv commented on build_run_docker.sh:
```diff
@@ -347,58 +333,43 @@ AUTOBENCH_DBT2_TEMPLATE=$(cat ./resources/config_templates/autobench_dbt2.conf)
 
-source $SCRIPT_DIR/docker.env
-source $SCRIPT_DIR/misc_configs.env
+# shellcheck source=./docker.env
```
You want to keep this here?
Also, what if the script is executed from somewhere else, does it still work?

svenssonaxel replied:
Yes, it's a linter directive, and is needed to not get warnings about missing variable declarations when they are found in the `.env` files. It's inside a comment so won't affect execution.

olapiv commented on build_run_docker.sh:
```diff
@@ -278,11 +257,10 @@ RONDB_VERSION_NO_DOT=$(echo "$RONDB_VERSION" | tr -d '.')
 FILE_SUFFIX="v${RONDB_VERSION_NO_DOT}_m${NUM_MGM_NODES}_g${NODE_GROUPS}_r${REPLICATION_FACTOR}_my${NUM_MYSQL_NODES}_api${NUM_API_NODES}"
 
 AUTOGENERATED_FILES_DIR="$SCRIPT_DIR/autogenerated_files/$FILE_SUFFIX"
-rm -rf $AUTOGENERATED_FILES_DIR
```
Why not just delete it all? Are you expecting to restart a cluster with previous data?

svenssonaxel replied:
This came along by mistake when I reverted some of your changes. I'll put it back in.

olapiv commented on build_run_docker.sh:
```diff
@@ -347,58 +333,43 @@ AUTOBENCH_DBT2_TEMPLATE=$(cat ./resources/config_templates/autobench_dbt2.conf)
+        target: %s"
 
 ENV_FIELD="
       environment:"
```
You could just add `HOST_GROUP_ID` here since all containers should have it.

svenssonaxel replied:
Done

olapiv commented on build_run_docker.sh:
```diff
@@ -416,10 +387,10 @@ VOLUMES=()
+    local IS_NDBD_DATA_DIR="$3"
+    if [ "$VOLUME_TYPE" == local ]; then
         local VOLUME_DIR="$LOCAL_VOLUMES_DIR/$VOLUME_NAME"
+        rm -rf "$VOLUME_DIR"
```
Might as well just delete all autogenerated files in the beginning?

svenssonaxel replied:
Yup.

olapiv reviewed (changes requested):
I'll test it now, but so far some comments..!

olapiv commented on build_run_docker.sh:
```diff
@@ -746,7 +724,12 @@ else
+# Make directories accessible by group members. This is used as a workaround
+# since the local (host) user likely has a different UID from the user in the
+# docker container.
+chmod g=u -R "$AUTOGENERATED_FILES_DIR"
```
This returns `chmod: -R: No such file or directory` on Mac

olapiv replied:
```bash
chmod --help           
chmod: illegal option -- -
usage:  chmod [-fhv] [-R [-H | -L | -P]] [-a | +a | =a  [i][# [ n]]] mode|entry file ...
        chmod [-fhv] [-R [-H | -L | -P]] [-E | -C | -N | -i | -I] file ...
```

svenssonaxel replied:
Fixed.

olapiv commented on resources/entrypoints/main.sh:
```diff
@@ -40,13 +57,38 @@ else
+			# failure.
+			true
+			sleep 5
+		done &
```
Btw, what happens if the ndbmtd crashes? Could be that the last files are not deletable anymore? Which PID does this even use if the ndbmtd is started with `exec`?

svenssonaxel replied:
> Btw, what happens if the ndbmtd crashes? Could be that the last files are not deletable anymore?

Yes, if ndbmtd exits within 5 seconds after creating a new file we're in trouble as you say. I'll fix it, but that means I can't use `exec` for `ndbmtd`.

> Which PID does this even use if the ndbmtd is started with `exec`?

The `&` after `done` creaes a sub shell, so the background bash process becomes a child of `ndbmtd` which looks kinda funny:
```
docker exec ndbd_1 bash -c 'ps -fuxa; grep -H Umask /proc/*/status'
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root        1477  0.0  0.0   4360  3176 ?        Ss   07:29   0:00 bash -c ps -fuxa; grep -H Umask /proc/*/status
root        1483  0.0  0.0   7060  1584 ?        R    07:29   0:00  \_ ps -fuxa
root           1  0.0  0.0   4360  3300 ?        Ss   07:02   0:00 /bin/bash ./docker_entrypoints/rondb_standalone/entrypoint.sh ndbmtd --ndb-nodeid=1 --initial --ndb-connectstring=mgmd_1:1186
root          30  0.0  0.0   7236  4348 ?        S    07:02   0:00 sudo -E -u mysql /srv/hops/docker_entrypoints/rondb_standalone/main.sh ndbmtd --ndb-nodeid=1 --initial --ndb-connectstring=mgmd_1:1186
mysql         31  0.0  0.0  29636 10216 ?        S    07:02   0:00  \_ ndbmtd --ndb-nodeid=1 --initial --ndb-connectstring=mgmd_1:1186 --nodaemon
mysql         34  0.0  0.0   4360  1776 ?        S    07:02   0:00      \_ /bin/bash /srv/hops/docker_entrypoints/rondb_standalone/main.sh ndbmtd --ndb-nodeid=1 --initial --ndb-connectstring=mgmd_1:1186
mysql       1476  0.0  0.0   2788  1008 ?        S    07:29   0:00      |   \_ sleep 5
mysql         40 29.3  8.0 6409960 5309384 ?     Sl   07:02   8:06      \_ ndbmtd --ndb-nodeid=1 --initial --ndb-connectstring=mgmd_1:1186 --nodaemon
/proc/1/status:Umask:   0022
/proc/1476/status:Umask:        0002
/proc/1477/status:Umask:        0022
/proc/30/status:Umask:  0022
/proc/31/status:Umask:  0002
/proc/34/status:Umask:  0002
/proc/40/status:Umask:  0002
/proc/self/status:Umask:        0022
/proc/thread-self/status:Umask: 0022

```

svenssonaxel replied:
But of course this will also not be the case after the fix, since I no longer use `exec`.

svenssonaxel committed:
Update sample files

svenssonaxel committed:
Fixes and small refactorings

build_run_docker.sh:
- Delete all of $AUTOGENERATED_FILES_DIR at startup.
- Refactor setting environment variables.
- Fix chmod call.

Entrypoint scripts:
- Fix debug prints.
- Fix group permissions in case of quick ndbmtd exit.

svenssonaxel commented:
All of the above now pushed.

olapiv commented on build_run_docker.sh:
```diff
@@ -624,41 +587,40 @@ if [ $NUM_API_NODES -gt 0 ]; then
+            add_file_to_template "$AUTOBENCH_SYS_SINGLE_FILEPATH" "$BENCH_DIR/sysbench_single/autobench.conf"
+            add_file_to_template "$AUTOBENCH_SYS_MULTI_FILEPATH" "$BENCH_DIR/sysbench_multi/autobench.conf"
+            add_file_to_template "$AUTOBENCH_DBT2_SINGLE_FILEPATH" "$BENCH_DIR/dbt2_single/autobench.conf"
+            add_file_to_template "$AUTOBENCH_DBT2_MULTI_FILEPATH" "$BENCH_DIR/dbt2_multi/autobench.conf"
```
This fails unless you use the `[ "$NUM_MYSQL_NODES" -gt 1 ]` check, such as this:

```
if [ "$VOLUME_TYPE" == "docker" ]; then
    add_file_to_template "$DBT2_CONF_SINGLE_FILEPATH" "$BENCH_DIR/dbt2_single/dbt2_run_1.conf"
    add_file_to_template "$AUTOBENCH_DBT2_SINGLE_FILEPATH" "$BENCH_DIR/dbt2_single/autobench.conf"
    add_file_to_template "$AUTOBENCH_SYS_SINGLE_FILEPATH" "$BENCH_DIR/sysbench_single/autobench.conf"
    if [ "$NUM_MYSQL_NODES" -gt 1 ]; then
        add_file_to_template "$DBT2_CONF_MULTI_FILEPATH" "$BENCH_DIR/dbt2_multi/dbt2_run_1.conf"
        add_file_to_template "$AUTOBENCH_DBT2_MULTI_FILEPATH" "$BENCH_DIR/dbt2_multi/autobench.conf"
        add_file_to_template "$AUTOBENCH_SYS_MULTI_FILEPATH" "$BENCH_DIR/sysbench_multi/autobench.conf"
    fi
fi
```

svenssonaxel replied:
Fixed, except I also added `if [ "$NUM_MYSQL_NODES" -gt 0 ]`

olapiv commented:
I just realised this also changes the API of the script - now, the absence of `-lv` affects all directories, also the benchmarking ones. There are parts in the README which assume that benchmarking directories are always mounted locally. For example:

> The results of the benchmarks are mounted into the local filesystem into the "autogenerated_files" directory. Look for "final_result.txt" in the directory of the benchmark that was run to see the results. For more information on how to read the benchmarking output, refer to the [RonDB documentation](http://docs.rondb.com) once again.

olapiv commented on resources/entrypoints/main.sh:
```diff
@@ -82,11 +82,23 @@ else
+		while true; do
+			ensure-group-permissions
 			sleep 5
 		done &
```
When running this, my terminal looks as follows:

```
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/1/T15F1.ctl' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/1/T15F0.ctl' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/1/T17F2.ctl' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/10/T17F1.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/10/T17F0.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/10/T17F3.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/10/T17F2.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/8/T17F1.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/8/T17F0.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/8/T17F3.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/8/T17F2.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/15/T17F1.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/15/T17F0.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/15/T17F3.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/15/T17F2.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/12/T17F1.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/12/T17F0.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/12/T17F3.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/12/T17F2.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/13/T17F1.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/13/T17F0.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/13/T17F3.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/13/T17F2.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/14/T17F1.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/14/T17F0.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/14/T17F3.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
ndbd_1    | mode of '/srv/hops/mysql-cluster/ndb_data/1/ndb_1_fs/LCP/14/T17F2.Data' changed from 0644 (rw-r--r--) to 0664 (rw-rw-r--)
```

Basically I can barely read anything else anymore. Is there any way of hiding this?

olapiv replied:
This happens a fair bit when benchmarks are run. And it is even run when omitting `-lv` (I don't mind that, I just mind the printing).

olapiv replied:
If possible maybe only print stderr

svenssonaxel replied:
I activated printing on purpose, but you're right that it's better not to print. I'll change it.

svenssonaxel replied:
Fixed.

svenssonaxel commented:
> I just realised this also changes the API of the script - now, the absence of `-lv` affects all directories, also the benchmarking ones. There are parts in the README which assume that benchmarking directories are always mounted locally.

Which route do you want to take here? I see several possibilities:
* Fix readme to explain `-lv` is needed to see benchmark results
* Change behavior so that benchmark dirs will always mount locally while the flag determines how to handle other volumes
* Remove the possibility of using docker volumes, and always use local mounts for everything.

I have a weak preference for the latter, but that's mostly a matter of taste. I feel like docker volumes are "out of control", or "cannot be owned", a little bit like global variables.

olapiv commented:
I think you'd also appreciate Docker volumes a bit more if you used Docker Desktop :) It's basically occupied storage that is difficult to forget, since you see it floating around there all the time.

So I'd just change the Readme.

svenssonaxel committed:
Fixes

- Fix readme to reflect that benchmark directories are only mounted if
  `-lv` is provided.
- In build_run_docker.sh, fix conditions for mounting files.
- Silence chmod.

svenssonaxel commented:
> change the Readme.

Fixed.
